File: scripts/check_repository.py
#!/usr/bin/env python3
"""Small, dependency-free repository policy checks used locally and in CI."""

from __future__ import annotations

import subprocess
import sys
import tomllib
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
REQUIRED_FILES = (
    "pyproject.toml",
    "uv.lock",
    ".python-version",
    ".node-version",
    ".pre-commit-config.yaml",
    ".github/workflows/ci.yml",
    ".github/dependabot.yml",
    "docs/development.md",
    "docs/github-protection.md",
    "CONTRIBUTING.md",
    "Makefile",
    "scripts/check_javascript.sh",
    "scripts/protect_main.sh",
)
FORBIDDEN_TRACKED_SUFFIXES = (".db", ".db-wal", ".db-shm", ".pem", ".key")
FORBIDDEN_TRACKED_NAMES = {".env", "coverage.xml", ".coverage"}


def tracked_files() -> list[str]:
    result = subprocess.run(
        ["git", "ls-files"],
        cwd=ROOT,
        check=False,
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        return []
    return [line.strip() for line in result.stdout.splitlines() if line.strip()]


def main() -> int:
    errors: list[str] = []

    for relative in REQUIRED_FILES:
        if not (ROOT / relative).is_file():
            errors.append(f"missing required workflow file: {relative}")

    pyproject_path = ROOT / "pyproject.toml"
    if pyproject_path.is_file():
        with pyproject_path.open("rb") as handle:
            project = tomllib.load(handle)
        if "project" not in project:
            errors.append("pyproject.toml has no [project] table")
        if "dev" not in project.get("dependency-groups", {}):
            errors.append("pyproject.toml has no dev dependency group")
        if "ruff" not in project.get("tool", {}):
            errors.append("pyproject.toml has no Ruff configuration")

    lock_path = ROOT / "uv.lock"
    if lock_path.is_file():
        if lock_path.stat().st_size < 500:
            errors.append("uv.lock looks incomplete")
        lock_text = lock_path.read_text(encoding="utf-8")
        for private_marker in ("applied-caas", "internal.api.openai.org"):
            if private_marker in lock_text:
                errors.append(
                    f"uv.lock contains a non-portable private package registry: {private_marker}"
                )
        if "https://pypi.org/simple" not in lock_text:
            errors.append("uv.lock is not resolved from the public PyPI index")

    for relative in tracked_files():
        path = Path(relative)
        if path.name in FORBIDDEN_TRACKED_NAMES:
            errors.append(f"forbidden generated/secret file is tracked: {relative}")
        if relative.startswith(".patch-backups/"):
            errors.append(f"patch backup is tracked: {relative}")
        if relative.startswith("backend/data/") and path.suffix in FORBIDDEN_TRACKED_SUFFIXES:
            errors.append(f"database file is tracked: {relative}")
        if path.suffix in {".pem", ".key"}:
            errors.append(f"possible private key is tracked: {relative}")

    workflow = ROOT / ".github/workflows/ci.yml"
    if workflow.is_file():
        text = workflow.read_text(encoding="utf-8")
        for required in ("pull_request:", "merge_group:", "name: CI Gate", "uv sync --frozen"):
            if required not in text:
                errors.append(f"CI workflow is missing: {required}")

    if errors:
        print("Repository policy checks failed:", file=sys.stderr)
        for error in errors:
            print(f"- {error}", file=sys.stderr)
        return 1

    print("Repository policy checks passed.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/check_repository.mjs
#!/usr/bin/env node

import { execFileSync } from "node:child_process"
import { readFileSync, statSync } from "node:fs"
import { fileURLToPath } from "node:url"
import { dirname, join, relative, resolve } from "node:path"

const root = resolve(dirname(fileURLToPath(import.meta.url)), "..")

const requiredFiles = [
  ".node-version",
  ".github/workflows/ci.yml",
  ".github/dependabot.yml",
  "CONTRIBUTING.md",
  "README.md",
  "package.json",
  "package-lock.json",
  "scripts/protect_main.sh",
  "supabase/config.toml",
  "wrangler.jsonc",
]

const retiredPaths = [
  ".patches/",
  "backend/",
  "docker/",
  "frontend/",
  "scripts/v01_",
]

const forbiddenNames = new Set([
  ".coverage",
  ".env",
  "coverage.xml",
])

const forbiddenSuffixes = [
  ".db",
  ".db-shm",
  ".db-wal",
  ".key",
  ".pem",
  ".sqlite",
  ".sqlite3",
]

function trackedFiles() {
  return execFileSync("git", ["ls-files"], {
    cwd: root,
    encoding: "utf8",
  })
    .split("\n")
    .filter(Boolean)
}

const errors = []

for (const path of requiredFiles) {
  try {
    if (!statSync(join(root, path)).isFile()) {
      errors.push(`missing required workflow file: ${path}`)
    }
  } catch {
    errors.push(`missing required workflow file: ${path}`)
  }
}

for (const path of trackedFiles()) {
  const name = path.split("/").at(-1)

  if (forbiddenNames.has(name)) {
    errors.push(`forbidden generated or secret file is tracked: ${path}`)
  }

  if (path.startsWith(".patch-backups/")) {
    errors.push(`patch backup is tracked: ${path}`)
  }

  if (retiredPaths.some((prefix) => path.startsWith(prefix))) {
    errors.push(`retired v0.1 path is tracked: ${path}`)
  }

  if (forbiddenSuffixes.some((suffix) => path.endsWith(suffix))) {
    errors.push(`database or possible private key is tracked: ${path}`)
  }
}

try {
  const lockText = readFileSync(join(root, "package-lock.json"), "utf8")
  const lock = JSON.parse(lockText)
  if (lock.lockfileVersion !== 3) {
    errors.push("package-lock.json must use lockfileVersion 3")
  }
} catch (error) {
  errors.push(`package-lock.json is invalid: ${error.message}`)
}

try {
  const workflowPath = join(root, ".github/workflows/ci.yml")
  const workflow = readFileSync(workflowPath, "utf8")
  for (const required of [
    "pull_request:",
    "merge_group:",
    "name: CI Gate",
    "npm run repository:check",
    "npm run lwin:test",
    "npm run web:ci",
    "npm run supabase -- test db",
  ]) {
    if (!workflow.includes(required)) {
      errors.push(`CI workflow is missing: ${required}`)
    }
  }
} catch (error) {
  errors.push(`unable to validate CI workflow: ${error.message}`)
}

if (errors.length > 0) {
  console.error("Repository policy checks failed:")
  for (const error of errors) {
    console.error(`- ${error}`)
  }
  process.exitCode = 1
} else {
  console.log(
    `Repository policy checks passed (${trackedFiles().length} tracked files from ${relative(process.cwd(), root) || "."}).`,
  )
}
